File: lib/pymodules/cmd_comm.py
################################################################################
#
# cmd_comm.c
#
# Various commands used in NakedMud(tm) for communicating with other
# characters, and NPCs.
#
################################################################################
from mud import *
from mudsys import add_cmd, add_cmd_check
import inform, hooks, mudsock, history



def cmd_ask(ch, cmd, arg):
    '''Usage: ask <person> [about] <question>

       This command is used to pose a question to another character. Mostly,
       this is intended to be used to carry on dialogs with NPCs. Ask has a
       local range (i.e. you can only ask questions to people in the same room
       as you.
       '''
    try:
        tgt, question = parse_args(ch, True, cmd, arg,
                                   "ch.room.noself [about] string(question)")
    except: return

    question = question.replace("$", "$$")
    message(ch, tgt, None, None, False, "to_vict",
            "{w$n asks you, '" + question + "'{n")
    message(ch, tgt, None, None, False, "to_char",
            "{wYou ask $N, '" + question + "'{n")

    # run our ask hooks
    hooks.run("ask", hooks.build_info("ch ch str", (ch, tgt, question)))

def cmd_tell(ch, cmd, arg):
    '''Usage: tell <person> <message>

       This command sends a message to another character. Primarily intended
       for player-to-player communication. Players can tell other players
       things even if they are not in the same room.

       see also: reply'''
    try:
        tgt, mssg = parse_args(ch, True, cmd, arg, "ch.world.noself string(message)")
    except: return

    mssg   = mssg.replace("$", "$$")
    tovict = "{r$n tells you, '" + mssg + "'{n"
    toch   = "{rYou tell $N, '" + mssg + "'{n"
    message(ch, tgt, None, None, False, "to_vict", tovict)
    message(ch, tgt, None, None, False, "to_char", toch)
    history.add_history(ch,   "tell", "{r%-10s: %s{n" % (ch.name, mssg))
    history.add_history(tgt,  "tell", "{r%-10s: %s{n" % (ch.name, mssg))
    hooks.run("tell", hooks.build_info("ch ch", (ch, tgt)))

def cmd_chat(ch, cmd, arg):
    '''Usage: chat <message>

       This command will send a message to all players currently logged on.
       '''
    if arg == '':
        ch.send("Chat what?")
    else:
        arg  = arg.replace("$", "$$")
        mssg = "{y$n chats, '" + arg + "'{n"
        message(ch, None, None, None, False, "to_world", mssg)
        message(ch, None, None, None,False,"to_char", "{yyou chat, '"+arg+"'{n")
        history.add_history(ch, "chat", "{y%-10s: %s{n" % (ch.name, arg))

def cmd_wiz(ch, cmd, arg):
    if arg == '':
        ch.send("WizChat what?")
    else:
        mssg = "{b%s WizChats, '{c%s{b'{n" % (ch.name, arg)
        for sock in mudsock.socket_list():
            if sock.ch != None and sock.ch.isInGroup("wizard"):
                sock.ch.send(mssg)

def cmd_say(ch, cmd, arg):
    '''Usage: say <message>

      This command will send a message to everyone in the same room as you. Say,
      like ask, can trigger NPC dialogs.'''
    if arg == '':
        ch.send("Say what?")
    else:
        arg = arg.replace("$", "$$")
        message(ch, None, None, None, False, "to_room",
                "{y$n says, '" + arg + "'{n")
        message(ch, None, None, None, False, "to_char",
                "{yyou say, '" + arg + "'{n")        

        # run say hooks
        hooks.run("say", hooks.build_info("ch str", (ch, arg)))

def cmd_greet(ch, cmd, arg):
    '''Usage: greet <person>

       NPCs with dialogs will often have something to say when you greet or
       approach then. Greeting an NPC is a way to get them talking.'''
    try:
        tgt, = parse_args(ch, True, cmd, arg, "ch.room.noself")
    except: return

    message(ch, tgt, None, None, False, "to_char", "You greet $N.")
    message(ch, tgt, None, None, False, "to_vict", "$n greets you.")
    message(ch, tgt, None, None, False, "to_room", "$n greets $N.")

    # run greet hooks
    hooks.run("greet",      hooks.build_info("ch ch", (ch, tgt)))
    hooks.run("post_greet", hooks.build_info("ch ch", (ch, tgt)))

def cmd_emote(ch, cmd, arg):
    '''Usage: emote <text>

       Send a special text message to the room you are in. The message is
       preceded by your name, unless you put a $n somewhere in the text, in
       which case the $n is replaced by your name. For example:

       > emote A gunshot sounds, and $n is laying on the ground, dead.

       Would show a message to everyone in the room saying that you are dead
       to a gunshot.'''
    if arg == '':
        ch.send("Emote we must, but emote what?")
    else:
        # see if a $n is within the argument ... if there is, let the person
        # put his or her name where it's wanted. Otherwise, tag it onto the
        # front of the message
        if arg.find("$n") == -1:
            arg = "$n " + arg
        message(ch, None, None, None, False, "to_room, to_char", arg)

def cmd_gemote(ch, cmd, arg):
    '''Gemote is similar to emote, except that it sends a mud-wide message
       instead of a room-specific message.'''
    if arg == '':
        ch.send("Gemote we must, but gemote what?")
    else:
        # same as emote, but global
        if arg.find("$n") == -1:
            arg = "$n " + arg
        message(ch, None, None, None, False, "to_world, to_char",
                "{bGLOBAL:{c " + arg + "{n")

def cmd_page(ch, cmd, arg):
    '''Usage: page <person> <message>

       Paging a person will send them a message, as well as making a beeping
       sound on their computer to get their attention. Page can be used on
       anyone in the mud, regardless if you are in the same room as them or not.
       '''
    try:
        tgt, mssg = parse_args(ch, True, cmd, arg, "ch.world.noself string(message)")
    except: return
    ch.send("\007\007You page " + ch.see_as(tgt))
    tgt.send("\007\007*" + tgt.see_as(ch) + "* " + mssg)



################################################################################
# add our commands
################################################################################
add_cmd("ask",     None, cmd_ask,   "player", False)
add_cmd("say",     None, cmd_say,   "player", False)
add_cmd("'",       None, cmd_say,   "player", False)
add_cmd("tell",    None, cmd_tell,  "player", False)
add_cmd("chat",    None, cmd_chat,  "player", False)
add_cmd("wizchat", "wiz",cmd_wiz,   "wizard", False)
add_cmd("gossip",  None, cmd_chat,  "player", False)
add_cmd("\"",      None, cmd_chat,  "player", False)
add_cmd("page",    None, cmd_page,  "player", False)
add_cmd("greet",   None, cmd_greet, "player", False)
add_cmd("approach",None, cmd_greet, "player", False)
add_cmd("emote",   None, cmd_emote, "player", False)
add_cmd("gemote",  None, cmd_gemote,"player", False)
add_cmd(":",       None, cmd_emote, "player", False)

def chk_room_communication(ch, cmd):
    if ch.pos in ("sleeping", "unconscious"):
        ch.send("You cannot do that while " + ch.pos + ".")
        return False

for cmd in ["ask", "say", "'", "greet", "approach", "emote", ":"]:
    add_cmd_check(cmd, chk_room_communication)

# register our history handling
history.register_comm_history("chat", lambda ch: "chat")
# history.register_comm_history("say",  lambda ch: ch.name)
history.register_comm_history("tell", lambda ch: ch.name)

File: src/cmd_comm.c
//*****************************************************************************
//
// cmd_comm.c
//
// Various commands used in NakedMud(tm) for communicating with other
// characters, and NPCs.
//
//*****************************************************************************

#include "mud.h"
#include "utils.h"
#include "handler.h"
#include "inform.h"
#include "character.h"
#include "world.h"
#include "dialog.h"


// option modules
#ifdef MODULE_SCRIPTS
#include "scripts/script.h"
#endif


//
// cmd_ask is used to pose a question to another character. Mostly, this is
// intended to be used to carry on dialogs with NPCs. Ask has a local range
// (i.e. you can only ask people in the same room as you questions)
//   usage: ask [person] <about> [question]
//
//   examples:
//     ask bob about cats           ask bob about the topic, "cats"
//     ask jim can I have a salad?  ask jim if you can have a salad
//
COMMAND(cmd_ask)
{
  if(!arg || !*arg)
    send_to_char(ch, "Ask who what?\r\n");
  else {
    bool ask_about = FALSE;
    char name[MAX_BUFFER];
    arg = one_arg(arg, name);
    trim(arg);

    // skip the "about" part in "ask [person] about"
    if(!strncasecmp(arg, "about ", 6)) {
      ask_about = TRUE;
      arg = arg+6;
    }

    if(!*arg)
      send_to_char(ch, "What did you want to ask %s?\r\n", name);
    else {
      CHAR_DATA *tgt = generic_find(ch, name,
				    FIND_TYPE_CHAR,
				    FIND_SCOPE_ROOM | FIND_SCOPE_VISIBLE,
				    FALSE, NULL);

      if(tgt == NULL)
	send_to_char(ch, "Who were you trying to ask a question?\r\n");
      else if(tgt == ch)
	send_to_char(ch, "You have a nice conversation with yourself.\r\n");
      else {
	char other_buf[MAX_BUFFER];
	sprintf(other_buf, "{w$n asks you%s, '%s'{n",
		(ask_about ? " about" : ""), arg);
	message(ch, tgt, NULL, NULL, FALSE, TO_VICT, other_buf);	

	send_to_char(ch, "{wYou ask %s%s, '%s'{n\r\n", 
		     charGetName(tgt), (ask_about ? " about" : ""), arg);

#ifdef MODULE_SCRIPTS
	try_speech_script(ch, tgt, arg);
#endif	

	// see if the NPC has something to say in return
	if(!try_dialog(ch, tgt, arg))
	  message(ch, tgt, NULL, NULL, TRUE, TO_CHAR,
		  "{p$N shrugs at the question.");
      }
    }
  }
}


//
// cmd_tell sends a message to another character. Primarily intended for
// player-player communication. Players can tell other players things even
// if they are not in the same room.
//   usage: tell [person] [mesage]
//
//   examples:
//     tell luke I am your father
//
COMMAND(cmd_tell)
{
  if(!arg || !*arg)
    send_to_char(ch, "Tell who what?\r\n");
  else {
    char name[MAX_BUFFER];
    arg = one_arg(arg, name);
    trim(arg);
    if(!*arg)
      send_to_char(ch, "What did you want to tell %s?\r\n", name);
    else {
      CHAR_DATA *tgt = generic_find(ch, name,
				    FIND_TYPE_CHAR,
				    FIND_SCOPE_WORLD | FIND_SCOPE_VISIBLE,
				    FALSE, NULL);

      if(tgt == NULL)
	send_to_char(ch, "Who were you trying to talk to?\r\n");
      else if(tgt == ch)
	send_to_char(ch, "You have a nice conversation with yourself.\r\n");
      else {
	send_to_char(ch, "{rYou tell %s, '%s'{n\r\n", charGetName(tgt), arg);

	// if we're an NPC, make sure we do colored output
	char *color_arg = arg;
	if(charIsNPC(ch)) {
	  DIALOG_DATA *dialog = worldGetDialog(gameworld, charGetDialog(tgt));
	  if(dialog)
	    color_arg = tagResponses(dialog, arg, "{c", "{r");
	}

	char other_buf[MAX_BUFFER];
	sprintf(other_buf, "{r$n tells you, '%s'{n", arg);
	message(ch, tgt, NULL, NULL, FALSE, TO_VICT, other_buf);

	if(color_arg != arg) 
	  free(color_arg);
      }
    }
  }
}


//
// cmd_chat sends a global message to all of the players currently logged on.
//   usage: chat [message]
//
//   example:
//     chat hello, world!
//
COMMAND(cmd_chat) {
  if (!arg || !*arg) {
    send_to_char(ch, "Chat what?\n\r");
    return;
  }
  communicate(ch, arg, COMM_GLOBAL);
}


//
// cmd_say sends a message to everyone in the same room as you. Say, like ask,
// can trigger NPC dialogs.
//   usage: say [message]
//
//   example:
//     say hello, room!
//
COMMAND(cmd_say) {
  if (!*arg) {
    send_to_char(ch, "Say what?\n\r");
    return;
  }

  char *color_arg = arg;
  if(charIsNPC(ch)) {
    DIALOG_DATA *dialog = worldGetDialog(gameworld, charGetDialog(ch));
    if(dialog)
      color_arg = tagResponses(dialog, arg, "{c", "{y");
  }

  communicate(ch, color_arg, COMM_LOCAL);
  if(color_arg != arg) free(color_arg);
}


//
// NPCs with dialogs will often have something to say when you greet/approach
// then. cmd_greet is a way to get them talking.
//   usage: greet [person]
//
//   examples:
//     greet mayor
//
COMMAND(cmd_greet) {
  if(!arg || !*arg)
    send_to_char(ch, "Whom did you want to greet?\r\n");
  else {
    CHAR_DATA *tgt = generic_find(ch, arg,
				  FIND_TYPE_CHAR,
				  FIND_SCOPE_ROOM | FIND_SCOPE_VISIBLE,
				  FALSE, NULL);

    if(tgt == NULL)
      send_to_char(ch, "Who were you trying to greet?\r\n");
    else if(tgt == ch)
      send_to_char(ch, 
		   "You shake your left hand with your right hand and say, "
		   "'nice to meet you, self'\r\n");
    else {
      message(ch, tgt, NULL, NULL, FALSE, TO_CHAR, 
	      "{wYou greet $N.");
      message(ch, tgt, NULL, NULL, FALSE, TO_VICT, 
	      "{w$n greets you.");
      message(ch, tgt, NULL, NULL, FALSE, TO_ROOM | TO_NOTCHAR | TO_NOTVICT, 
	      "$n greets $N.");
      
      // see if the NPC has something to say in return
      if(charIsNPC(tgt)) {
	DIALOG_DATA *dialog = worldGetDialog(gameworld, charGetDialog(tgt));
	
	// Do we have something to say, back?
	if(dialog && *dialogGetGreet(dialog)) {
	  char *response = tagResponses(dialog,
					dialogGetGreet(dialog),
					"{c", "{p");
	  send_to_char(ch, "{p%s responds, '%s'\r\n",charGetName(tgt),response);
	  free(response);
	}

	else
	  send_to_char(ch, "{p%s does not have anything to say.\r\n",
		       charGetName(tgt));
      }
    }
  }
}


//
// Send a special text message to the room you are in. The message is preceded
// by your name, unless you put a $n somewhere in the text, in which case the
// $n is replaced by your name.
//   usage: emote [message]
//
//   examples:
//     emote does a little dance.
//     emote A gunshot sounds, and $n is laying on the ground, dead.
//
COMMAND(cmd_emote) {
  if(!arg || !*arg)
    send_to_char(ch, "Emote we must, but emote what?\r\n");
  else {
    char buf[MAX_BUFFER];

    // see if a $n is within the argument ... if there is, let the
    // person put his or her name were it's wanted. Otherwise, tag
    // it onto the front of the message
    if(strfind(arg, "$n"))
      strcpy(buf, arg);
    else
      sprintf(buf, "$n %s", arg);

    message(ch, NULL, NULL, NULL, FALSE, TO_ROOM, buf);
  }
}


//
// cmd_gemote is similar to emote, but it sends a global message
//
COMMAND(cmd_gemote) {
  if(!arg || !*arg)
    send_to_char(ch, "Gemote we must, but gemote what?\r\n");
  else {
    char buf[MAX_BUFFER];
    if(strfind(arg, "$n"))
      sprintf(buf, "{bGLOBAL:{c %s", arg);
    else
      sprintf(buf, "{bGLOBAL:{c $n %s", arg);

    message(ch, NULL, NULL, NULL, FALSE, TO_WORLD, buf);
  }
}
